### ISSUE TITLE
Unable to Reproduce LLM2Vec Training Results Using GradCache on Echo Dataset

### ISSUE THREAD

**Issue by viet-data** [McGill-NLP/llm2vec#135]
I have been attempting to reproduce the training results on the same echo data. Due to hardware limitations, I had to reimplement the training process using GradCache.

Although my model code can load the LLM2Vec public checkpoint and perform inference correctly, I am unable to achieve comparable performance to LLM2Vec when training a bidirectional Mistral model (without MNTP and unsupervised SimCSE) using GradCache. My training used a batch size of 512 on the echo dataset and stopped after 750 iterations.

Specifically, on the STS tasks, I have not been able to exceed 75 on SICK-R and 65 on STS-12 (other tasks also show low performance, except for BIOSSES).

Has anyone else tried to train LLM2Vec with GradCache, or has anyone successfully reproduced the LLM2Vec results using the original code? Any insights or suggestions would be greatly appreciated.

**Comment by vaibhavad** [comment]
Hello @viet-data,

where you able to reproduce with GradCache? If you are interested, we'll like to integrate GradCache to the LLM2Vec library

**Comment by viet-data** [comment]
Hi @vaibhavad ,

I have successfully trained with gradcache, using a batch size of 128, and achieved results close to those reported in llm2vec. However, I'm curious about llm2vec's performance when scaling up the data. I haven't been able to improve performance with more training data, which might be due to the smaller batch size.

Could you share the llm2vec results when training on the full dataset? It would be very useful if you could integrate gradcache into llm2vec to help us train with fewer GPUs. Thank you.

**Comment by stefanhgm** [comment]
Hi @viet-data,

I reproduced the Llama 3 supervised version trained 1000 steps on the MNTP task and 1000 steps on the E5 dataset (according to the original LLM2Vec) training configs. I currently run the full MTEB evaluation, but the first results look very similar to the ones reported on HuggingFace for the model.

I am currently training a Llama 3.1 version with the same training recipe.

**Comment by viet-data** [comment]
@stefanhgm Thanks so much for sharing! I agree, llm2vec seems quite reproducible. Excited to see your results with Llama 3.1

**Comment by stefanhgm** [comment]
Currently, the evaluation of the Llama 3.1 version on MTEB hangs on a task where repeatedly 391 batches are processed. It already repeats this for over a day now. I think it is the `DBPedia` task as the `CQADupstackWordpressRetrieval` and `ClimateFEVER` tasks were completed last and `DBPedia` should come next.

@vaibhavad any chance you observed a similar behavior when evaluating on MTEB?

<img width="408" alt="Screenshot 2024-08-13 at 14 51 19" src="https://github.com/user-attachments/assets/e653972a-8888-494b-a82c-7e3a86cae74f">

**Comment by atutej** [comment]
Hi @stefanhgm!

Would it be possible to share your reproduced numbers? If I understand correctly, performance matches when you train with SimCSE followed by Supervised?

Currently I am training from MNTP as given in the training config files with the same hyper parameters, but my numbers are off for some benchmarks (eg: 48% versus 55% reported on FiQA.

**Comment by atutej** [comment]
Hi @stefanhgm!

Would it be possible to share your reproduced numbers? I am currently following the LLM2Vec recipe and for some benchmarks (eg: FiQA2018) the numbers I get are way off from what was reported. 48% vs 55%...

**Comment by stefanhgm** [comment]
Hi @atutej,

I still have problem with running all tasks because the running time is just very long even when using multiple GPUs. I also asked a question regarding there here: #140

I got FiQA2018 running though and obtained `"main_score": 0.55441` on the test set, so for me the results seem reproducible.

**Comment by vaibhavad** [comment]
> Currently, the evaluation of the Llama 3.1 version on MTEB hangs on a task where repeatedly 391 batches are processed. It already repeats this for over a day now. I think it is the `DBPedia` task as the `CQADupstackWordpressRetrieval` and `ClimateFEVER` tasks were completed last and `DBPedia` should come next.
> 
> @vaibhavad any chance you observed a similar behavior when evaluating on MTEB?
> 
> <img alt="Screenshot 2024-08-13 at 14 51 19" width="408" src="https://private-user-images.githubusercontent.com/3011151/357416134-e653972a-8888-494b-a82c-7e3a86cae74f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4OTU3ODEsIm5iZiI6MTcyNzg5NTQ4MSwicGF0aCI6Ii8zMDExMTUxLzM1NzQxNjEzNC1lNjUzOTcyYS04ODg4LTQ5NGItYTgyYy03ZTNhODZjYWU3NGYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDJUMTg1ODAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mzk4NDk5NzZhNjI5NTlkMjAxMDcyODRlMTUwYzY2NWM2ZGI2ZmM5OWUyOTg0ZjUzZGM5NTA0OWVhZjg2MThkYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.vv_-pNBRveE41H8T9oxJleTqJZfVA_lOHpoaECeHiHI">

I haven't faces this issue. Can you share you evaluation script?

**Comment by atutej** [comment]
Hi @stefanhgm 

I think the 391 batches are just sub-batches in the dataset. I see the same thing but it eventually finishes evaluating. 

Regarding training: are you doing both mntp followed by supervised? I'm starting from mntp checkpoint provided on huggingface for supervised training. @vaibhavad is it possible there are some differences between following either of the above methods?

**Comment by stefanhgm** [comment]
Hi @atutej 

I trained mntp from scratch and supervised after it. So, I did not use the version from HuggingFace.

Here is my mteb eval code #140
